server/auth/registrar.go

```diff
@@ -25,7 +30,344 @@ var (
+		return msgjson.NewError(msgjson.BondError, "invalid bond transaction: %v", err)
+	}
+	if amt < int64(auth.minBond) {
+		return msgjson.NewError(msgjson.BondError, "insufficient bond amount %d, needed %d", amt, auth.minBond)
+	}
+	if lockTime < lockTimeThresh.Unix() {
+		return msgjson.NewError(msgjson.BondError, "insufficient lock time %d, needed at least %d", lockTime, lockTimeThresh.Unix())
+	}
+
+	// Must be equal to account ID computed from pubkey in the PayFee message.
+	if commitAcct != acctID {
+		return msgjson.NewError(msgjson.BondError, "invalid bond transaction - account commitment does not match pubkey")
+	}
+
+	// Verify that the user has the private keys to the bond pubkey by verifying
+	// the BondSig against the message (acct ID) and bond script's pubkey hash.
+	msgHash := sha256.Sum256(acctID[:])
+	sigPK, _, err := ecdsa.RecoverCompact(postBond.BondSig, msgHash[:])
+	if err != nil {
+		return msgjson.NewError(msgjson.BondError, "invalid bond signature: %v", err)
+	}
+	if !bytes.Equal(bondPK, sigPK.SerializeCompressed()) {
+		return msgjson.NewError(msgjson.BondError, "bond signature does not match bond pubkey")
+	}
+
+	// All good. The client gets a PostBondResult (no error) unless the confirms
+	// check has an unexpected error or times out.
+	expireTime := time.Unix(lockTime, 0).Add(-auth.bondExpiry)
+	postBondRes := &msgjson.PostBondResult{
+		AccountID: acctID[:],
+		AssetID:   assetID,
+		Amt:       uint64(amt),
+		Expiry:    expireTime.Unix(),
+		BondID:    bondCoinID,
+		// Confs is set below.
+	}
+	auth.Sign(postBondRes)
+
+	sendResp := func() *msgjson.Error {
+		resp, err := msgjson.NewResponse(msg.ID, postBondRes, nil)
+		if err != nil { // shouldn't be possible
+			return msgjson.NewError(msgjson.RPCInternalError, "internal encoding error")
+		}
+		err = conn.Send(resp)
+		if err != nil {
+			log.Warnf("Error sending postbond result to user %v: %v", acctID, err)
+			if err = auth.Send(acctID, resp); err != nil {
+				log.Warnf("Error sending postbond result to account %v: %v", acctID, err)
+				// The user will need to 'connect' to reconcile bond status.
+			}
+		}
+		return nil
+	}
+
+	// See if the account exists, and get known unexpired bonds. Also see if the
+	// account has previously paid a legacy registration fee.
+	dbAcct, bonds, _, legacyFeePaid := auth.storage.Account(acctID, lockTimeThresh)
+	newAcct := dbAcct == nil
+
+	reqConfs := int64(auth.bondConfs)
+
+	// See if we already have this bond in DB.
+	for _, bond := range bonds {
+		if bond.AssetID == assetID && bytes.Equal(bond.CoinID, bondCoinID) {
+			postBondRes.Active = true
+			_, postBondRes.Tier = auth.AcctStatus(acctID)
+			return sendResp()
+		}
+	}
+
+	bondStr := coinIDString(assetID, bondCoinID)
+	bondAssetSym := dex.BipIDSymbol(assetID)
+
+	// Determine if it exists and is confirmed.
+	chkAddr, chkAmt, confs, err := auth.checkFee(assetID, bondCoinID)
+	if err != nil {
+		// Not found is expected if this is a pre-broadcast validation request.
```

Comment:
Alternative to always checking and interpreting a coin-not-found to mean it's a pre-validation request, the request payload could simply indicate this intention.